### DISCUSSION TITLE
Handling Different Database Names in Partitions and Git Branching

### DISCUSSION THREAD

**Discussion by atanasov-hristo** [TabularEditor/TabularEditor3#1414] [🙋♀️TE3 Q&A]
Hello! We are trying to migrate from Visual Studio to Tabular Editor 3 and we are starting to use the “Save to Folder” feature which enables us to utilize Git branches for our tabular model project.

The problem that we are running into now is that even though the single Model.bim file is now split into many smaller JSON files; we still have a partition file for each table where the name of the database is part of the expression.

We have different physical databases for our Production and Stage environments. Right now, we have two separate tabular models where the partition expression has references to each particular database. Since we are trying to utilize a single Git repository for out project with “master” and “stage” branches, for example; those differences in the partition’s expression would cause problems upon switching between our “master”, “stage” and topic branches and them merging changes between branches.

My question is what are the best practices to handle the different database names in partitions in different branches for each environment (e.g. PROD vs STAGE) in the same Git repo?

**Comment by otykier** [comment]
I see a couple of issues that needs to be addressed here:

1) **Redundancy in M expressions.** Having the server/database name specified in every M expression across all partitions is error prone and makes it difficult to retarget the model to use a different database when needed. The solution to this is to create a **Shared Expression**. Either one that contains the `Sql.Database` function call, or two shared expressions returning strings - one for the server name, and another for the database name.
2) **Configuration stored in git** Server/database names that are part of the model metadata should be treated as *configuration*. It's fine to have placeholder values in the model metadata (which may point to the dev/stage environment), but you should not rely on branches to provide, say, the prod environment values. This is similar to how you would never store credentials (i.e. usernames and passwords) in git. Instead, you should substitute the correct values as part of your *deployment* of the model. For example, with Azure DevOps, it's possible to have different environment variables that gets swapped in depending on whether you're deploying to stage/prod, etc. For example, you can swap in these variables using a simple PowerShell script, or you can do it using the TE2 CLI and a C# script. Also, if you implement the shared expressions as mentioned in point 1) above, the replacement operation becomes a lot simpler, as you don't have to perform the replace across all the partitions in the model.

Hope this helps!
